# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib.colors as colors
import numpy as np
print(os.getcwd())

D:\Springboard_DataSci\Assignments\Capstone_1


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
path = r'D:\Springboard_DataSci\Assignments\Capstone_1\data'
os.chdir(path)
df = pd.read_csv('step3_output.csv')
df = df.drop('Unnamed: 0', axis=1)
print(df.head)

<bound method NDFrame.head of                                   Name       state  vertical_drop  base_elev  \
0                     Hilltop Ski Area      Alaska            294       1796   
1                  Sunrise Park Resort     Arizona           1800       9200   
2        Yosemite Ski & Snowboard Area  California            600       7200   
3               Boreal Mountain Resort  California            500       7200   
4                          Dodge Ridge  California           1600       6600   
..                                 ...         ...            ...        ...   
167                      Hogadon Basin     Wyoming            640       7400   
168          Sleeping Giant Ski Resort     Wyoming            810       6619   
169                   Snow King Resort     Wyoming           1571       6237   
170  Snowy Range Ski & Recreation Area     Wyoming            990       8798   
171                White Pine Ski Area     Wyoming           1100       8400   

     fast

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
df = df.assign(state_id=df['state'].astype('category').cat.codes)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [4]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','state','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [5]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [6]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [7]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [8]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model
from sklearn import metrics

In [9]:
print('Explained variance score:', metrics.explained_variance_score(y_test, y_pred))
print('Mean absolute error:', metrics.mean_absolute_error(y_test, y_pred))

Explained variance score: 0.7802835545827156
Mean absolute error: 6.498558303877332


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [10]:
print('Intercept:', lm.intercept_)

Intercept: 56.42414758443106


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [11]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(coef.sort_values(by='Coefficient', ascending=False)[:10])

                   Coefficient
AdultWeekday          9.366030
vertical_drop         3.146329
daysOpenLastYear      2.973156
projectedDaysOpen     2.337383
SkiableTerrain_ac     1.926559
triple                1.664099
LongestRun_mi         1.577742
quad                  1.377764
surface               1.014006
base_elev             0.943781


**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [12]:
X_no_state = X.drop(['state_id'], axis=1)
scaler = preprocessing.StandardScaler().fit(X_no_state)
X_scaled=scaler.transform(X_no_state)

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [13]:
print('Explained variance score:', metrics.explained_variance_score(y_test, y_pred))
print('Mean absolute error:', metrics.mean_absolute_error(y_test, y_pred))
print('Intercept:', lm.intercept_)
coef = pd.DataFrame(abs(lm.coef_), X_no_state.columns, columns=['Coefficient'])
print(coef.sort_values(by='Coefficient', ascending=False)[:10])

Explained variance score: 0.782058255769979
Mean absolute error: 6.451301377820562
Intercept: 56.43413814227478
                   Coefficient
AdultWeekday          9.412102
vertical_drop         3.258159
daysOpenLastYear      3.035465
projectedDaysOpen     2.455452
SkiableTerrain_ac     2.042277
triple                1.591136
LongestRun_mi         1.571052
quad                  1.401396
surface               1.011677
base_elev             0.971241


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [14]:
X_less_elev = X.drop(['base_elev','state_id'], axis=1)
scaler = preprocessing.StandardScaler().fit(X_less_elev)
X_scaled=scaler.transform(X_less_elev)

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
model = lm.fit(X_train,y_train)
y_pred = model.predict(X_test)

In [15]:
print('Explained variance score:', metrics.explained_variance_score(y_test, y_pred))
print('Mean absolute error:', metrics.mean_absolute_error(y_test, y_pred))
print('Intercept:', lm.intercept_)
coef = pd.DataFrame(abs(lm.coef_), X_less_elev.columns, columns=['Coefficient'])
print(coef.sort_values(by='Coefficient', ascending=False)[:10])

Explained variance score: 0.781771892467625
Mean absolute error: 6.482144939327319
Intercept: 56.44978531173757
                   Coefficient
AdultWeekday          9.361340
vertical_drop         3.275352
daysOpenLastYear      2.922757
projectedDaysOpen     2.413939
SkiableTerrain_ac     2.050665
LongestRun_mi         1.605731
triple                1.546328
quad                  1.383355
clusters              1.255846
surface               0.941904


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.7802 | 6.498 |-|
| Model 2. | 0.7820| 6.451 |'state_id'|
| Model 3. | 0.769 | 6.507 |'base_elev','state_id'|

Model Selection:

Model 3. Three features were removed, and the loss of explained variance was minimal. A ski resort cannot change its base elevation without relocating, so it makes little since to include it as a feature.